User defined Function
 ┌──────────────────────────────┐
 │      User define function     │
 │ def raj(x):                  │
 │     return x['Name']         │
 └──────────────┬───────────────┘
                │
                │  You call it
                ▼
 ┌──────────────────────────────┐
 │        raj(df)               │
 └──────────────┬───────────────┘
                │
                ▼
        x  =  df   (df replaces x)
                │
                ▼
 ┌──────────────────────────────┐
 │  return x['Name']            │
 │  → return df['Name']         │
 └──────────────┬───────────────┘
                │
                ▼
     Output: df['Name'] column (Series)

In short

x is just a placeholder for whatever you pass.

When you call raj(df), Python internally does:
x = df
return x['Name']
→ which means → return df['Name'].

Lambda function
raj = lambda x: x['Name']
raj → is the name of the function (just like def raj(x): ...).

x → is the parameter (a placeholder).

When we call raj(df), our DataFrame df replaces x.

In [1]:
def add(*x):
    addition = 0

    for i in x:
        addition = addition + i
    return addition

In [2]:
add(3,5,4)

12

In [3]:
# using the same wth lambda
new_add = lambda *x: sum(x)
new_add(12,10)

22

In [4]:
def ad(x):
    result = sum(x)
    return result

In [5]:
raj = [12,12]
ad(raj)

24

In [6]:
# using the same with lambda
new_ad = lambda x: sum(x)
new_ad(raj)

24

In [7]:
def sub(a,b):
    return a - b

In [8]:
sub(10,12)

-2

In [9]:
# using the same with lambda

new_sub = lambda a,b: a-b
new_sub(10,15)

-5

In [10]:
# doubling a number

double = lambda x: x*2
double(2)

4

In [11]:
# raise it to power 2
pow = lambda x: x**2
pow(3)

9

In [12]:
# finding percentage of numbers is a list

listb = [10, 20, 30]
per = lambda x: [round((i *100)/sum(x), 2) for i in listb]
per(listb)

[16.67, 33.33, 50.0]

In [13]:
# average
lista = [2,4,5,6,7,8]
average = lambda x: sum(x)/len(x)
average(lista)

5.333333333333333

In [14]:
dict1 = {'Name': ['Raj', 'Ratnajit', 'Reshmi', 'Rohan', 'Sreya'],
         'Math_Scores': [98, 99, 35, 38, 56],
         'Gender': ['Male', 'Male', 'Female', 'Male', 'Female']}

import pandas as pd
df = pd.DataFrame(dict1)
print(df)

       Name  Math_Scores  Gender
0       Raj           98    Male
1  Ratnajit           99    Male
2    Reshmi           35  Female
3     Rohan           38    Male
4     Sreya           56  Female


In [15]:
import re
df['Gender'] =  df['Gender'].apply(lambda x: re.sub('Female', 'Male', x))
df

Name  Math_Scores Gender
0       Raj           98   Male
1  Ratnajit           99   Male
2    Reshmi           35   Male
3     Rohan           38   Male
4     Sreya           56   Male

df['Gender'] = df['Gender'].apply(lambda x: re.sub('Female', 'Male', x))

Step	Value of x	After re.sub('Female', 'Male', x)
1	'Female'	'Male'
2	'Male'	'Male'
3	'Female'	'Male'
4	'Female'	'Male'

in this line
df['Gender'].apply(lambda x: ...)
x = one single value (one cell) from the column df['Gender'].

In [16]:
# taking two parameters
avg = lambda x, y: (x+y)/2
avg(3,2)

2.5

In [17]:
vizon = lambda *x: sum(x)/len(x)
vizon(5,10,10)

8.333333333333334

function inside funbction

In [18]:
cube = lambda x: x*x*x

In [19]:
def appl(x, z):
    return 6 + x(z)

In [20]:
appl(cube, 2)

14

# Higher order function

In [21]:
avg = lambda x, y: (x+y)/2
cube = lambda x: x*x*x

def ken(avg, x, y, cube):
    return avg(x,y) - cube(x)


In [22]:
ken(avg, 2,3,cube)

-5.5

Create a function that:

Takes two numbers

Takes another function (like add, subtract, multiply, divide) as an argument

Applies that function dynamically

In [23]:
add = lambda x,y: x+y
sub = lambda x,y: x - y
mutiply = lambda x,y: x*y
div = lambda x,y: x/y

def fun(operation,x,y):
    return operation(x,y)

In [24]:
fun(add,10,13)

23

Build a function that:

Takes a list of numbers

Takes a list of transformation functions (like square, cube, double, etc.)

Applies each transformation to the data and returns all results

In [25]:
square = lambda x: x*x
cube = lambda x: x * x * x
double = lambda x: x*2

def appl(operation, x):
    return operation(x)

In [26]:
appl(double, 5)

10

In [27]:
import pandas as pd
data = [
    {"Name": "Amit", "Marks": 85},
    {"Name": "Riya", "Marks": 92},
    {"Name": "Soham", "Marks": 76},
    {"Name": "Neha", "Marks": 89},
    {"Name": "Raj", "Marks": 95}
]

df = pd.DataFrame(data)
print(df)

    Name  Marks
0   Amit     85
1   Riya     92
2  Soham     76
3   Neha     89
4    Raj     95


In [28]:
import numpy as np
maximum = lambda data: df.loc[df[df['Marks'] == df['Marks'].max()].index, 'Name']
add_category = lambda data: data.assign(Category=np.where(data['Marks'] < 92, 'Bad', np.where(data['Marks'] < 80, 'Fail', 'Pass')))

def create(operation, data):
    return operation(data)

In [29]:
create(add_category, df)

Name  Marks Category
0   Amit     85      Bad
1   Riya     92     Pass
2  Soham     76      Bad
3   Neha     89      Bad
4    Raj     95     Pass

In [30]:
with_5 = lambda x: x* 5
with_10 = lambda x: x* 10

def doit(z, x):
    return z(x)

In [31]:
doit(with_10, 10)

100

# combining lambda filter & map

In [32]:
listc = [100, 200, 300]
result = list(filter(lambda x: x < 100, map(lambda x: x/50, listc)))
print(result)

[2.0, 4.0, 6.0]


In [33]:
df

Name  Marks
0   Amit     85
1   Riya     92
2  Soham     76
3   Neha     89
4    Raj     95

In [34]:
def assi(data):
    category = [data['Marks'] > 90, (data['Marks'] >=80) & (data['Marks'] < 90),
                (data['Marks'] >=70) & (data['Marks'] < 80),
                data['Marks'] < 70]
    assign_grade = ['A', 'B', 'C', 'D']
    data['Grade'] = np.select(category, assign_grade, default='E')
    return data

In [35]:
assi(df)

Name  Marks Grade
0   Amit     85     B
1   Riya     92     A
2  Soham     76     C
3   Neha     89     B
4    Raj     95     A

In [36]:
help(np.select)

Help on _ArrayFunctionDispatcher in module numpy:

select(condlist, choicelist, default=0)
    Return an array drawn from elements in choicelist, depending on conditions.

    Parameters
    ----------
    condlist : list of bool ndarrays
        The list of conditions which determine from which array in `choicelist`
        the output elements are taken. When multiple conditions are satisfied,
        the first one encountered in `condlist` is used.
    choicelist : list of ndarrays
        The list of arrays from which the output elements are taken. It has
        to be of the same length as `condlist`.
    default : scalar, optional
        The element inserted in `output` when all conditions evaluate to False.

    Returns
    -------
    output : ndarray
        The output at position m is the m-th element of the array in
        `choicelist` where the m-th element of the corresponding array in
        `condlist` is True.

    See Also
    --------
    where : Return elements from

| Condition                         | Rating      |
| --------------------------------- | ----------- |
| Sales > 80 and CSAT > 90          | ⭐ Excellent |
| Sales > 70 and CSAT between 80–90 | 👍 Good     |
| Sales > 50 and CSAT between 70–80 | 😐 Average  |
| Anything else                     | 👎 Poor     |


In [37]:
import pandas as pd

# Creating the dataset
data = {
    'Name': ['Amit', 'Riya', 'Soham', 'Neha', 'Raj', 'Tina', 'Karan', 'Leena', 'Arjun', 'Pooja'],
    'Sales': [85, 75, 60, 40, 92, 55, 78, 67, 88, 50],
    'Customer_Satisfaction': [95, 82, 75, 60, 88, 73, 89, 79, 91, 65]
}

df = pd.DataFrame(data)
print(df)


    Name  Sales  Customer_Satisfaction
0   Amit     85                     95
1   Riya     75                     82
2  Soham     60                     75
3   Neha     40                     60
4    Raj     92                     88
5   Tina     55                     73
6  Karan     78                     89
7  Leena     67                     79
8  Arjun     88                     91
9  Pooja     50                     65


In [38]:
import numpy as np
def assi(data):
    condition = [(data['Sales'] > 80) & (data['Customer_Satisfaction'] > 90),
                 (data['Sales'] > 70) & (data['Customer_Satisfaction'] >= 80) & (data['Customer_Satisfaction'] <=90),
                 (data['Sales'] > 50) & (data['Customer_Satisfaction'] >=70) & (data['Customer_Satisfaction'] <=80)]
    category = ['⭐Excellent', '👍 Good', '😐 Average']
    data['Level'] = np.select(condition, category, default='👎 Poor')
    return data

In [39]:
assi(df)

Name  Sales  Customer_Satisfaction       Level
0   Amit     85                     95  ⭐Excellent
1   Riya     75                     82      👍 Good
2  Soham     60                     75   😐 Average
3   Neha     40                     60      👎 Poor
4    Raj     92                     88      👍 Good
5   Tina     55                     73   😐 Average
6  Karan     78                     89      👍 Good
7  Leena     67                     79   😐 Average
8  Arjun     88                     91  ⭐Excellent
9  Pooja     50                     65      👎 Poor

In [40]:
# lets do the same with lambda
categorization = lambda data: data.assign(Level=np.select([(data['Sales'] > 80) & (data['Customer_Satisfaction'] > 90),
                 (data['Sales'] > 70) & (data['Customer_Satisfaction'] >= 80) & (data['Customer_Satisfaction'] <=90),
                 (data['Sales'] > 50) & (data['Customer_Satisfaction'] >=70) & (data['Customer_Satisfaction'] <=80)],
                                         ['⭐Excellent', '👍 Good', '😐 Average'], default='👎 Poor'))


categorization(df)                                 

Name  Sales  Customer_Satisfaction       Level
0   Amit     85                     95  ⭐Excellent
1   Riya     75                     82      👍 Good
2  Soham     60                     75   😐 Average
3   Neha     40                     60      👎 Poor
4    Raj     92                     88      👍 Good
5   Tina     55                     73   😐 Average
6  Karan     78                     89      👍 Good
7  Leena     67                     79   😐 Average
8  Arjun     88                     91  ⭐Excellent
9  Pooja     50                     65      👎 Poor

# Higher Order function
Goal:

Create a Higher-Order Function that can apply different operations to a DataFrame — like cleaning, grading, or labeling — by accepting another function as input.

You’ll pass different lambda functions or custom functions into it.

1. import pandas as pd
df = pd.DataFrame({
    'Name': ['Amit', 'Riya', 'Soham', 'Neha', 'Raj'],
    'Marks': [85, 92, 76, 69, 95],
    'Attendance': [88, 92, 70, 60, 96]
})

2.Write a higher-order function called process() that:

Takes two arguments:

a DataFrame

an operation function (that will be applied to the DataFrame)

Returns the processed DataFrame.
def process(data, operation):
    return operation(data)

3.Now, create three different lambda or normal functions you can pass to process():

1.grade_students: Adds a Grade column using np.select()

2. attendance_status: Adds a Status column (Good, Low)

3. name_uppercase: Converts all student names to uppercase

In [41]:
import pandas as pd
df = pd.DataFrame({
    'Name': ['Amit', 'Riya', 'Soham', 'Neha', 'Raj'],
    'Marks': [85, 92, 76, 69, 95],
    'Attendance': [88, 92, 70, 60, 96]
})
df

Name  Marks  Attendance
0   Amit     85          88
1   Riya     92          92
2  Soham     76          70
3   Neha     69          60
4    Raj     95          96

In [42]:
import numpy as np
grading = lambda data: data.assign(Grade=np.select([data['Marks'] > 90, data['Marks'] > 80, data['Marks'] > 70],
                                         ['A', 'B', 'C'], default='D'))

give_status = lambda data: data.assign(status=np.select([data['Attendance'] > 90, data['Attendance'] > 80],
                                         ['Good', 'Low'], default='Low'))

upper_case = lambda data: data.assign(Name=data['Name'].str.upper())

def store(operation, data):
    return operation(data)



In [43]:
store(grading,df)

Name  Marks  Attendance Grade
0   Amit     85          88     B
1   Riya     92          92     A
2  Soham     76          70     C
3   Neha     69          60     D
4    Raj     95          96     A

In [44]:
store(give_status, df)

Name  Marks  Attendance status
0   Amit     85          88    Low
1   Riya     92          92   Good
2  Soham     76          70    Low
3   Neha     69          60    Low
4    Raj     95          96   Good

In [45]:
store(upper_case, df)

Name  Marks  Attendance
0   AMIT     85          88
1   RIYA     92          92
2  SOHAM     76          70
3   NEHA     69          60
4    RAJ     95          96

In [46]:
# using lambda and filter - Direct expression
lista = [2,3,10,16,15,22,454]

even = list(filter(lambda x: x % 2 == 0, lista))
print(even)

[2, 10, 16, 22, 454]


In [47]:
#lambda version
even = lambda z: list(filter(lambda x: x % 2 == 0, z))
lista = [2,3,10,16,15,22,454]
even(lista)

[2, 10, 16, 22, 454]

🔹 Task:

Write a lambda function that takes a list of integers and:

Doubles each number using map.

Filters only the numbers greater than 10 using filter.

Sorts the result in descending order using sorted + key=lambda.

method_chaining = lambda z: list(filter(lambda x: x > 10, map(lambda x: x * 2, filter(lambda x: isinstance(x, int), z))))
listab = ['2', 4, 89, 'Ratnajit', 'Python', 89]
method_chaining(listab)

In [48]:
# replace and lambda
name = lambda z: list(map(lambda x: x.replace(x, 'Raj'), filter(lambda x: len(x) == 3, z)))
lista = ['Raja', 'Dee', 'Roh']
name(lista)

['Raj', 'Raj']

In [49]:
dict1 = {'Name': ['Raj', 'Ratnajit', 'Reshmi', 'Rohan', 'Sreya'],
         'Math_Scores': [98, 99, 35, 38, 56],
         'Gender': ['Male', 'Male', 'Female', 'Male', 'Female']}

import pandas as pd
df = pd.DataFrame(dict1)
print(df)

       Name  Math_Scores  Gender
0       Raj           98    Male
1  Ratnajit           99    Male
2    Reshmi           35  Female
3     Rohan           38    Male
4     Sreya           56  Female


In [50]:
#replacing name - use of apply
import re
replace = lambda x: x['Name'].apply(lambda x: re.sub('Raj', 'Ratnajit', x))
replace(df)

0    Ratnajit
1    Ratnajit
2      Reshmi
3       Rohan
4       Sreya
Name: Name, dtype: object

In [51]:
# assign use
df.assign(Updated_name = df['Name'].apply(lambda x: re.sub('Raj', 'Ratnajit', x)))
df


Name  Math_Scores  Gender
0       Raj           98    Male
1  Ratnajit           99    Male
2    Reshmi           35  Female
3     Rohan           38    Male
4     Sreya           56  Female

.apply() → 'Go inside the column and modify each cell'

.assign() → 'Add or update a column based on some logic'


🧠 Difference Between .apply() and .assign() in Pandas

🔹 .apply() → Works inside a column

Used to apply a function to each value in a Series (column) or row.

It transforms individual cells/values.

Usually used with lambda, regex, or math operations.
ex:
df['Name'] = df['Name'].apply(lambda x: re.sub('Raj', 'Ratnajit', x))


🔹 .assign() → Works at the DataFrame level

Used to add or update entire columns based on logic.

Always returns a new DataFrame (does not modify the original unless reassigned).

Often combined with .apply() to build new or cleaned columns.
ex:
df = df.assign(Updated_Name=df['Name'].apply(lambda x: re.sub('Raj', 'Ratnajit', x)))


| Purpose                                  | Use                                      |
| ---------------------------------------- | ---------------------------------------- |
| Modify values inside a column            | `.apply()`                               |
| Add or update a column                   | `.assign()`                              |
| Combine transformation + column creation | `.assign(column = df['col'].apply(...))` |


df['Gender'] = df['Gender'].apply(lambda x: re.sub('Female', 'Male', x))

Step	Value of x	After re.sub('Female', 'Male', x)
1	'Female'	'Male'
2	'Male'	'Male'
3	'Female'	'Male'
4	'Female'	'Male'

in this line
df['Gender'].apply(lambda x: ...)
x = one single value (one cell) from the column df['Gender'].

In [52]:
df.apply(lambda x: x['Name'].upper(), axis=1)
df

Name  Math_Scores  Gender
0       Raj           98    Male
1  Ratnajit           99    Male
2    Reshmi           35  Female
3     Rohan           38    Male
4     Sreya           56  Female

In [69]:
import pandas as pd
# Sample DataFrame
df5 = pd.DataFrame({
    'Math': [90, 85, 70],
    'Science': [80, 75, 95],
    'English': [88, 92, 85]
})
print(df5)


   Math  Science  English
0    90       80       88
1    85       75       92
2    70       95       85


In [70]:
# Apply a function column-wise (axis=0)
result = df5.apply(lambda x: x.std() - x.mean(), axis=0)
print(result)

Math      -71.258337
Science   -72.925003
English   -84.821449
dtype: float64


here lamba x is thw whole columns since axis=0
Maths
x = [90, 85, 70]
x.std()  = 10.408
x.mean() = 81.67
x.std() - x.mean() = -71.26

Science
x = [80, 75, 95]
x.std()  = 10.408
x.mean() = 83.33
x.std() - x.mean() = -72.92

English
x = [88, 92, 85]
x.std()  = 3.511
x.mean() = 88.33
x.std() - x.mean() = -84.82


🧠 Understanding lambda x: in apply()

When you use df.apply(lambda x: ..., axis=1):

👉 x represents one complete row of the DataFrame (as a Pandas Series).

The function runs once per row, so you can access individual columns inside that row using x['column_name'].
In simple words .apply() will loop through each row (since axis=1).

📘 Example:
df.apply(lambda x: x['Name'].upper(), axis=1)


Here, x is one row at a time.

x['Name'] gives the Name value from that specific row.

🔍 Example Illustration:

For a DataFrame:

Name	Math_Scores	Gender
Raj	98	Male
Reshmi	35	Female

The lambda function runs like this:

For the first row →
x = {'Name': 'Raj', 'Math_Scores': 98, 'Gender': 'Male'}

For the second row →
x = {'Name': 'Reshmi', 'Math_Scores': 35, 'Gender': 'Female'}

✅ In short:
When using apply(..., axis=1),
➡️ x = one row
When using apply(..., axis=0),
➡️ x = one column

# Lambda and llist comprehension

In [53]:
dict1 = {'Name': ['Raj', 'Ratnajit', 'Reshmi', 'Rohan', 'Sreya'],
         'Math_Scores': [98, 99, 35, 38, 56],
         'Gender': ['Male', 'Male', 'Female', 'Male', 'Female']}

import pandas as pd
df = pd.DataFrame(dict1)
print(df)

       Name  Math_Scores  Gender
0       Raj           98    Male
1  Ratnajit           99    Male
2    Reshmi           35  Female
3     Rohan           38    Male
4     Sreya           56  Female


In [54]:
df['Name'] = df['Name'].apply(lambda x: [i.upper() for i in x])
df

Name  Math_Scores  Gender
0                 [R, A, J]           98    Male
1  [R, A, T, N, A, J, I, T]           99    Male
2        [R, E, S, H, M, I]           35  Female
3           [R, O, H, A, N]           38    Male
4           [S, R, E, Y, A]           56  Female

Here in the above case lambda x, x is a single value from a column so x['Name'] is a value from name column,x = Raj, x =Ratnajit

In [55]:
dict1 = {'Name': ['Raj', 'Ratnajit', 'Reshmi', 'Rohan', 'Sreya'],
         'Math_Scores': [98, 99, 35, 38, 56],
         'Gender': ['Male', 'Male', 'Female', 'Male', 'Female']}

import pandas as pd
df1 = pd.DataFrame(dict1)
print(df1)

       Name  Math_Scores  Gender
0       Raj           98    Male
1  Ratnajit           99    Male
2    Reshmi           35  Female
3     Rohan           38    Male
4     Sreya           56  Female


In [56]:
standard_deviation = lambda x: x['Math_Scores'].std()

In [57]:
standard_deviation(df1)

31.443600302764313

Here in the above example x is the whole df x['Math_scores'] means df1['Math_Scores']

In [58]:
# lambda and list comprehension
upper_list = lambda x: [i.upper() for i in x if str(i).isalpha()]
lista = [12,45,'Raj']
upper_list(lista)

['RAJ']

here lambda x, x is the whole list
           ┌────────────────────────────┐
           │ upper_list(lista) called   │
           └────────────┬───────────────┘
                        │
                        ▼
               ┌─────────────────────┐
               │  lambda x: ...      │
               │  (x = lista)        │
               └─────────┬───────────┘
                         │
                         ▼
          x = [12, 45, 'Raj']   ←── whole list passed as x
                         │
                         ▼
         ┌──────────────────────────────────┐
         │ for i in x → loop through each i │
         └──────────────────────────────────┘
            │             │              │
            ▼             ▼              ▼
          i=12          i=45           i='Raj'
            │             │              │
       str(i).isalpha()?  ✗              ✓
            │                             │
            ├───────────────→─────────────┤
            ▼                             ▼
   ignored (not alphabetic)     str('Raj').upper() = 'RAJ'

                         │
                         ▼
               ┌──────────────────┐
               │ ['RAJ'] returned │
               └──────────────────┘


# .Map() pandas function

In [59]:
import pandas as pd

# Sample DataFrame
df = pd.DataFrame({
    'Name': ['Ravi', 'Sreya', 'Ankit', 'Priya', 'Rekha'],
    'Age': [23, 25, 22, 24, 26],
    'City': ['Delhi', 'Kolkata', 'Mumbai', 'Chennai', 'Pune'],
    'Gender': ['Male', 'Female', 'Male', 'Female', 'Female']
})

print(df)

    Name  Age     City  Gender
0   Ravi   23    Delhi    Male
1  Sreya   25  Kolkata  Female
2  Ankit   22   Mumbai    Male
3  Priya   24  Chennai  Female
4  Rekha   26     Pune  Female


In [60]:
df['Name'] = df['Name'].map(lambda x: x.upper() if x.startswith('R') else x)
df

Name  Age     City  Gender
0   RAVI   23    Delhi    Male
1  Sreya   25  Kolkata  Female
2  Ankit   22   Mumbai    Male
3  Priya   24  Chennai  Female
4  REKHA   26     Pune  Female

here lambda x, x is a single value just like when we apply df['col'].apply(lambda x)
same with .map
Here .map is a pandas function which automatically loops (iterates) over each element internally.
So here df['Name'].map, here map looks over each value in Name columns one by one and apply .upper function

In [61]:
import pandas as pd

# Sample DataFrame
df3 = pd.DataFrame({
    'Name': ['Ravi', 'Sreya', 'Ankit', 'Priya', 'Rekha'],
    'Age': [23, 25, 22, 24, 26],
    'City': ['Delhi', 'Kolkata', 'Mumbai', 'Chennai', 'Pune'],
    'Gender': ['Male', 'Female', 'Male', 'Female', 'Female']
})

print(df3)

    Name  Age     City  Gender
0   Ravi   23    Delhi    Male
1  Sreya   25  Kolkata  Female
2  Ankit   22   Mumbai    Male
3  Priya   24  Chennai  Female
4  Rekha   26     Pune  Female


In [62]:
df3['Name'] = df3['Name'].map(lambda x: 'Mr/Mr.s ' + x)
df3

Name  Age     City  Gender
0   Mr/Mr.s Ravi   23    Delhi    Male
1  Mr/Mr.s Sreya   25  Kolkata  Female
2  Mr/Mr.s Ankit   22   Mumbai    Male
3  Mr/Mr.s Priya   24  Chennai  Female
4  Mr/Mr.s Rekha   26     Pune  Female

In [63]:
# df['col'].apply is also same as map. it also loops over each value in that column
#reverse
df3['Name'] = df3['Name'].apply(lambda x: x[::-1])
df3

Name  Age     City  Gender
0   ivaR s.rM/rM   23    Delhi    Male
1  ayerS s.rM/rM   25  Kolkata  Female
2  tiknA s.rM/rM   22   Mumbai    Male
3  ayirP s.rM/rM   24  Chennai  Female
4  ahkeR s.rM/rM   26     Pune  Female